fe/fe-core/src/main/java/org/apache/doris/clone/DynamicPartitionScheduler.java

```diff
@@ -140,6 +142,69 @@ private Map<String, String> createDefaultRuntimeInfo() {
         return defaultRuntimeInfo;
     }
 
+    // exponential moving average
+    private static long ema(ArrayList<Long> history, int period) {
+        double alpha = 2.0 / (period + 1);
+        double ema = history.get(0);
+        for (int i = 1; i < history.size(); i++) {
+            ema = alpha * history.get(i) + (1 - alpha) * ema;
+        }
+        return (long) ema;
+    }
+
+    private static long getNextPartitionSize(ArrayList<Long> historyPartitionsSize) {
+        if (historyPartitionsSize.size() < 2) {
+            return historyPartitionsSize.get(0);
+        }
+
+        int size = historyPartitionsSize.size() > 7 ? 7 : historyPartitionsSize.size();
+
+        boolean isAscending = true;
+        for (int i = 1; i < size; i++) {
+            if (historyPartitionsSize.get(i) < historyPartitionsSize.get(i - 1)) {
+                isAscending = false;
+                break;
+            }
+        }
+
+        if (isAscending) {
+            ArrayList<Long> historyDeltaSize = Lists.newArrayList();
+            for (int i = 1; i < size; i++) {
+                historyDeltaSize.add(historyPartitionsSize.get(i) - historyPartitionsSize.get(i - 1));
+            }
+            return historyPartitionsSize.get(size - 1) + ema(historyDeltaSize, 7);
+        } else {
+            return ema(historyPartitionsSize, 7);
+        }
+    }
+
+    private static int getBucketsNum(DynamicPartitionProperty property, OlapTable table) {
+        if (!table.isAutoBucket()) {
+            return property.getBuckets();
+        }
+
+        // auto bucket
+        List<Partition> partitions = Lists.newArrayList(table.getPartitions());
+        if (partitions.size() == 0) {
+            return property.getBuckets();
+        }
+
+        Collections.sort(partitions, new Comparator<Partition>() {
+            @Override
+            public int compare(Partition p1, Partition p2) {
+                return (int) (p1.getId() - p2.getId());
```

Comment:
You can not rely on partition id to get the latest partition.
Use partition range value instead.

Reply:
And looks like you sort the `partitions` but not use it?

Reply:
Sort `partitions` for get last n history partition size